lib/db_connection/holder.ex

```diff
@@ -46,30 +65,106 @@ defmodule DBConnection.Holder do
   end
 
-  @spec stop(pool_ref :: any, err :: Exception.t(), state :: any) :: :ok
-  def stop(pool_ref, err, state) do
-    done(pool_ref, :stop, state, err)
+  @spec stop(pool_ref :: any, err :: Exception.t()) :: :ok
+  def stop(pool_ref, err) do
+    done(pool_ref, {:error, err}, :stop, err)
   end
 
-  @spec update(pid, reference, module, term) :: t
-  def update(pool, ref, mod, state) do
-    holder = new(pool, ref, mod, state)
-    now = System.monotonic_time(@time_unit)
-    :ets.give_away(holder, pool, {:checkin, ref, now})
-    holder
+  @spec handle(pool_ref :: any, fun :: atom, args :: [term], Keyword.t) :: tuple
+  def handle(pool_ref, fun, args, opts) do
+    handle(:handle, pool_ref, fun, args, opts)
+  end
+
+  @spec cleanup(pool_ref :: any, fun :: atom, args :: [term], Keyword.t) :: tuple
+  def cleanup(pool_ref, fun, args, opts) do
+    handle(:cleanup, pool_ref, fun, args, opts)
+  end
+
+  defp handle(type, pool_ref, fun, args, opts) do
+    pool_ref(holder: holder) = pool_ref
+
+    try do
+      :ets.lookup(holder, :conn)
+    rescue
+      ArgumentError ->
+        msg = "connection is closed"
+        {:disconnect, DBConnection.ConnectionError.exception(msg), _state = :unused}
+    else
+      [conn(status: {:error, _})] ->
+        msg = "connection is closed"
+        {:disconnect, DBConnection.ConnectionError.exception(msg), _state = :unused}
+
+      [conn(status: :aborted)] when type != :cleanup ->
+        msg = "transaction rolling back"
+        {:disconnect, DBConnection.ConnectionError.exception(msg), _state = :unused}
+
+      [conn(module: module, state: state)] ->
+        try do
+          apply(module, fun, args ++ [opts, state])
+        catch
+          kind, reason ->
+            {:catch, kind, reason, System.stacktrace()}
+        else
+          result when is_tuple(result) ->
+            state = :erlang.element(:erlang.tuple_size(result), result)
+
+            # This means a disconnect happened from the connection side
+            # but, since we succeed, we can return the result and the client
+            # will notice the disconnect anyway.
+            try do
+              :ets.update_element(holder, :conn, {conn(:state) + 1, state})
+            rescue
+              ArgumentError -> false
+            end
+
+            result
+
+          # If it is not a tuple, we just return it as is so we raise bad return.
+          result ->
+            result
+        end
+    end
+  end
+
+  ## Pool state helpers API (invoked by callers)
+
+  @spec put_state(pool_ref :: any, term) :: :ok
+  def put_state(pool_ref(holder: sink_holder), state) do
+    :ets.update_element(sink_holder, :conn, [{conn(:state) + 1, state}])
+    :ok
+  end
+
+  @spec status?(pool_ref :: any, :ok | :aborted) :: boolean()
+  def status?(pool_ref(holder: holder), status) do
+    try do
+      :ets.lookup_element(holder, :conn, conn(:status) + 1) == status
+    rescue
+      ArgumentError -> false
```

Comment:
does false make sense if `status` is aborted? See earlier comment so may not matter. Seems like an additional status.

Reply:
In my mind, the missing ETS entry is equivalent to the pdict being removed. And all of the times we call `Holder.status?`we are not interested in those cases, hence false is reasonable.